services/venice-controller/src/main/java/com/linkedin/venice/controller/init/ControllerClientBackedSystemSchemaInitializer.java

```diff
@@ -0,0 +1,246 @@
+      }
+    } catch (VeniceNoStoreException e) {
+      checkAndMayCreateSystemStore(storeName, schemasInLocalResources.get(1));
+    }
+
+    // Only verify the key schema if it is explicitly specified by the caller. We don't care about the dummy key schema.
+    if (keySchema != null) {
+      checkIfKeySchemaMatches(storeName);
+    }
+
+    MultiSchemaResponse multiSchemaResponse =
+        controllerClient.retryableRequest(DEFAULT_RETRY_TIMES, c -> c.getAllValueSchema(storeName));
+    if (multiSchemaResponse.isError()) {
+      throw new VeniceException(
+          "Error when getting all value schemas from system store " + storeName + " in cluster " + clusterName
+              + " after retries. Error: " + multiSchemaResponse.getError());
+    }
+    Map<Integer, Schema> schemasInZk = new HashMap<>();
+    Arrays.stream(multiSchemaResponse.getSchemas())
+        .forEach(schema -> schemasInZk.put(schema.getId(), AvroCompatibilityHelper.parse(schema.getSchemaStr())));
+    for (int version = 1; version <= protocolDefinition.getCurrentProtocolVersion(); version++) {
+      Schema schemaInLocalResources = schemasInLocalResources.get(version);
+      if (schemaInLocalResources == null) {
+        throw new VeniceException(
+            "Invalid protocol definition: " + protocolDefinition.name() + " does not have a version " + version
+                + " even though that is inferior to the current version ("
+                + protocolDefinition.getCurrentProtocolVersion() + ").");
+      }
+      checkAndMayRegisterValueSchema(storeName, version, schemasInZk.get(version), schemaInLocalResources);
+
+      if (autoRegisterDerivedComputeSchema) {
+        checkAndMayRegisterWriteComputeSchema(storeName, version, schemaInLocalResources);
+      }
+    }
+  }
+
+  private void checkAndMayCreateSystemStore(String storeName, Schema firstValueSchema) {
+    StoreResponse storeResponse = controllerClient.retryableRequest(
+        DEFAULT_RETRY_TIMES,
+        c -> c.getStore(storeName),
+        r -> r.getError().contains("does not exist"));
+    if (storeResponse.isError()) {
+      if (storeResponse.getError().contains("does not exist")) {
+        if (firstValueSchema == null) {
+          throw new VeniceException("Protocol definition: " + protocolDefinition.name() + " does not have version 1");
+        }
+        String firstKeySchemaStr = keySchema == null ? DEFAULT_KEY_SCHEMA_STR : keySchema.toString();
+        String firstValueSchemaStr = firstValueSchema.toString();
+        NewStoreResponse newStoreResponse = controllerClient.retryableRequest(
+            DEFAULT_RETRY_TIMES,
+            c -> c.createNewSystemStore(
+                storeName,
+                VeniceConstants.SYSTEM_STORE_OWNER,
+                firstKeySchemaStr,
+                firstValueSchemaStr),
+            r -> r.getError().contains("already exists"));
+        if (newStoreResponse.isError() && !newStoreResponse.getError().contains("already exists")) {
```

Comment:
Yeah unfortunately current APIs sometimes do not respond with accurate error type. For example, when retrieving a non-existent store, error type is GENERAL_ERROR instead of STORE_NOT_FOUND. So I have to leverage error message to handle store not exist/already exist cases for now.